samples/demos/end_to_end/main.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2020, Digi International, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from hdc1080 import HDC1080
+from machine import I2C
+import time
+import xbee
+
+# Constants
+SLEEP_TIME_MS = 8000  # 8 seconds.
+AWAKE_TIME_MS = 2000  # 2 seconds.
+
+MSG_ON = "ON"
+MSG_OFF = "OFF"
+MSG_SEPARATOR = "@@@"
+MSG_AWAKE = "AWAKE"
+
+print(" +------------------------------------+")
+print(" | XBee MicroPython End to End Sample |")
+print(" +------------------------------------+\n")
+
+# Instantiate the HDC1080 peripheral.
+sensor = HDC1080(I2C(1))
+
+# Instantiate the XBee device.
+xb = xbee.XBee()
+
+# Configure sleep mode to be managed by MicroPython.
+xb.atcmd("SM", 0x06)
+
+sleep_time = SLEEP_TIME_MS
+
+# Start reading temperature and humidity measures.
+while True:
```

Comment:
How would you read team and humidity as well as poll messages from mqtt? Im stuck as the while True blocks the main thread.

Reply:
You could check for new messages using the `check_msg()` method inside the loop (supposing that you are using our umqtt library and you have connected the client and registered the appropriate callback):

```python
while True:
    # Check for pending messages from the server. If there is any, the registered callback will be invoked.
    c.check_msg()
    # Notify the gateway that the XBee is awake.
    ...
```

Reply:
Thanks you!